### PR TITLE
Make H2 query improvements common [DPP-600]

### PR DESCRIPTION
**Pull request by kamil-da**
### Motivation
This is a follow-up PR to #10888 . It ports H2-specific changes for all db types.

The aim is to start work on removing `participant_events` view as it proved to have negative impact on performance for H2 and Oracle.

### Testing
I've run the 800tps `sandbox-on-x` benchmark and observed no performance degradation.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/QueryStrategy.scala [hunk 1]
```diff
@@ -55,4 +55,6 @@ trait QueryStrategy {
 
   /** Boolean predicate */
   def isTrue(booleanColumnName: String): String
+
+  def nullAs(typeName: String): String = s"NULL::$typeName"
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/common/QueryStrategy.scala:59]: Two comments:
- the double colon operator is not standard, I would not provide a default implementation
- type names are not portable - see for example `BIGINT` vs `NUMBER`

How about adding a `def nullAsNumber: string` instead? Then you wouldn't need the overrideable `NullAsLedgerEffectiveTimeType`.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/ContractStorageBackendTemplate.scala [hunk 2]
```diff
@@ -100,10 +64,45 @@ trait ContractStorageBackendTemplate extends ContractStorageBackend {
     if (ids.isEmpty) {
       Failure(emptyContractIds)
     } else {
-      def lookup(id: ContractId): Option[Option[Instant]] =
-        maximumLedgerTimeSqlLiteral(id).as(instantFromMicros("ledger_effective_time").?.singleOpt)(
+      def lookup(id: ContractId): Option[Option[Instant]] = {
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/common/ContractStorageBackendTemplate.scala:67]: Why did you inline `maximumLedgerTimeSqlLiteral` here? Could you have just modified `maximumLedgerTimeSqlLiteral` instead?

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/oracle/OracleStorageBackend.scala [hunk 3]
```diff
@@ -116,9 +116,10 @@ private[backend] object OracleStorageBackend
     override def arrayIntersectionNonEmptyClause(
         columnName: String,
         parties: Set[Ref.Party],
-    ): CompositeSql =
-      cSQL"EXISTS (SELECT 1 FROM JSON_TABLE(#$columnName, '$$[*]' columns (value PATH '$$')) WHERE value IN (${parties
-        .map(_.toString)}))"
+    ): CompositeSql = {
+      val inParties = parties.map(p => s"'$p'").mkString(", ")
+      cSQL"EXISTS (SELECT 1 FROM JSON_TABLE(#$columnName, '$$[*]' columns (value PATH '$$')) WHERE value IN (#$inParties))"
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/backend/oracle/OracleStorageBackend.scala:121]: Why this change? If I understand correctly, instead of using a query parameter, parties are now hardcoded in the SQL string.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/ContractStorageBackendTemplate.scala (hunk 2 above)
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/ContractStorageBackendTemplate.scala:67]: The `maximumLedgerTimeSqlLiteral` was introduced in my previous PR #10888 in order to create a different implementation for H2. Currently there's a single implementation, so I reverted the `lookup()` method to have an inlined query.

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/oracle/OracleStorageBackend.scala (hunk 3 above)
Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/oracle/OracleStorageBackend.scala:121]: This is the fix for the issue with SQL query syntax error on Oracle. I still need to figure out the exact difference in rendering comparing to the proposed version. Without this change the active contract query lookup query fails on Oracle.

Comment (meiersi-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/oracle/OracleStorageBackend.scala:121]: If this value is really inlined then that potentially opens the query up to SQL injection attacks. Let's make sure that we pass the values as a parameter to the query.

Comment (kamil-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/oracle/OracleStorageBackend.scala:121]: I'm on it!

**Comment by kamil-da** [comment]
Removing the `participant_events` view revealed a possible bug in Oracle when using the `JSON_TABLE` function on a `CLOB` column, prepared statements and an index on the same column at the same time.
We're going to postpone removing the view due to some upcoming work that is going to facilitate or possibly cover the task anyway.
